### PR TITLE
WT-10349 Implement list_objects() in connection class for Azure

### PR DESCRIPTION
**Pull request by mariammojid**
Implemented the list_objects functionality for azure connection using list_blobs azure sdk function.

### CODE REVIEW

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 1]
```diff
@@ -31,16 +31,38 @@
...
+
 azure_connection::azure_connection(const std::string &bucket_name, const std::string &obj_prefix)
     : _azure_client(Azure::Storage::Blobs::BlobContainerClient::CreateFromConnectionString(
         std::getenv("AZURE_STORAGE_CONNECTION_STRING"), bucket_name)),
       _bucket_name(bucket_name), _object_prefix(obj_prefix)
 {
 }
 
+/*
+ * Lists all of the objects (blobs) on the bucket (container) utilising the listBlobs function form
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/azure_connection.cpp:47]: Love the commenting, I would reword this more of: "Build a list of all of the objects from the bucket." We should describe the concept of blobs and container either the constructor or the .h file.

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 2]
```diff
@@ -31,16 +31,38 @@
...
+ * the Azure SDK
+ */
 int
-azure_connection::list_objects(std::vector<std::string> &objects) const
+azure_connection::list_objects(
+  const std::string &prefix, std::vector<std::string> &objects, bool list_single) const
 {
+    objects::clear();
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/azure_connection.cpp:54]: Why do we clear the objects vector here?

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 3]
```diff
@@ -31,16 +31,38 @@
...
+    
+    Azure::Storage::Blobs::ListBlobsOptions blob_parameters;
+    blob_parameters.Prefix = prefix;
+
+    auto list_blobs_response = _azure_client.ListBlobs(blob_parameters);
+
+    for (auto blob_item : list_blobs_response.Blobs) {
+        objects.push_back(blob_item.Name);
+        if (list_single)
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/azure_connection.cpp:63]: Could we pass in a parameter to list the number of objects that we can list? If so, we can remove this if statement.

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 4]
```diff
@@ -31,16 +31,38 @@
...
+ */
 int
-azure_connection::list_objects(std::vector<std::string> &objects) const
+azure_connection::list_objects(
+  const std::string &prefix, std::vector<std::string> &objects, bool list_single) const
 {
+    objects::clear();
+    
+    Azure::Storage::Blobs::ListBlobsOptions blob_parameters;
+    blob_parameters.Prefix = prefix;
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/azure_connection.cpp:57]: Have you tested this functionality?

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 5]
```diff
@@ -31,16 +31,38 @@
...
+
+    auto list_blobs_response = _azure_client.ListBlobs(blob_parameters);
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/azure_connection.cpp:59]: What happens if list_blobs_response fails? Is that possible?

File: ext/storage_sources/azure_store/test/test_azure_connection.cpp [hunk 6]
```diff
@@ -29,6 +29,29 @@
 #define CATCH_CONFIG_MAIN
 #include <catch2/catch.hpp>
 
+#include <stdlib.h>
+#include <iostream>
+
 #include "azure_connection.h"
 
-TEST_CASE("testing class azure_connection", "azure_connection") {}
+// Includes necessary for connection
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/test/test_azure_connection.cpp:37]: I would revert this file, since we don't want commented code in production.

**Review by jiechenbo — COMMENTED** [review]
Please add a detailed comment describing what was done in this ticket on JIRA.

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 1 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp:47]: ok I've added a comment before the contructor:
Buckets are synonymous with Containers, and Objects are synonymous with Blobs.

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 2 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp:54]: Initially, we cleared the object vector when doing testing but the function assumes that the user gives in an empty objects vector

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 3 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp:63]: I've replaced the if condition and accounted for the list_single in the blob_parameters: 
```
   if (list_single)
    blob_parameters.PageSizeHint = 1;
```
Where [pagesizehint](https://azuresdkdocs.blob.core.windows.net/$web/cpp/azure-storage-blobs/12.0.0/struct_azure_1_1_storage_1_1_blobs_1_1_list_blobs_options.html) indicates the maximum number of blobs to return

File: ext/storage_sources/azure_store/test/test_azure_connection.cpp (hunk 6 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/test/test_azure_connection.cpp:37]: reverted :)

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 4 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp:57]: yes we've tested this functionality, ran them on Lachlan's machine with the put_objects function
image

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 5 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp:59]: nope it doesnt fail
https://azuresdkdocs.blob.core.windows.net/$web/cpp/azure-storage-blobs/12.0.0/class_azure_1_1_storage_1_1_blobs_1_1_list_blobs_paged_response.html

File: ext/storage_sources/azure_store/test/test_azure_connection.cpp (hunk 6 above)
Comment (jiechenbo) [reply ext/storage_sources/azure_store/test/test_azure_connection.cpp:37]: It doesn't seem to be reverted it yet :(

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 7]
```diff
@@ -31,16 +31,35 @@
...
+
 azure_connection::azure_connection(const std::string &bucket_name, const std::string &obj_prefix)
     : _azure_client(Azure::Storage::Blobs::BlobContainerClient::CreateFromConnectionString(
         std::getenv("AZURE_STORAGE_CONNECTION_STRING"), bucket_name)),
       _bucket_name(bucket_name), _object_prefix(obj_prefix)
 {
 }
 
+/*
+ * Build a list of all of the objects in the bucket
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/azure_connection.cpp:47]: full stop at end of sentence.

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 8]
```diff
@@ -31,16 +31,35 @@
...
+ */
 int
-azure_connection::list_objects(std::vector<std::string> &objects) const
+azure_connection::list_objects(
+  const std::string &prefix, std::vector<std::string> &objects, bool list_single) const
 {
+    Azure::Storage::Blobs::ListBlobsOptions blob_parameters;
+    blob_parameters.Prefix = prefix;
+    if (list_single)
+        blob_parameters.PageSizeHint = 1;
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/azure_connection.cpp:56]: Have we tested this?

File: ext/storage_sources/azure_store/azure_connection.h [hunk 9]
```diff
@@ -33,10 +33,14 @@
 
 #include <string>
 
+/*
+ * In Azure: Buckets are synonymous with Containers, and Objects are synonymous with Blobs.
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/azure_connection.h:37]: If we are planning to add a comment here, we should add a description of what the class does as a whole. Can you follow the s3_connection.h comment and describe what azure connection class should be doing?

File: ext/storage_sources/azure_store/test/test_azure_connection.cpp (hunk 6 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/test/test_azure_connection.cpp:37]: srry mb reverted

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 7 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp]: added :)

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 8 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp:56]: yes :) passed unit tests we ran on lachlans machine

File: ext/storage_sources/azure_store/azure_connection.h (hunk 9 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.h:37]: comment added now :)

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 10]
```diff
@@ -31,16 +31,35 @@
 #include <azure/core.hpp>
 #include <azure/storage/blobs.hpp>
 
+#include <stdlib.h>
```
Comment (korteland) [inline ext/storage_sources/azure_store/azure_connection.cpp:34]: `#include <cstdlib>` is more idiomatic for C++.

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 11]
```diff
@@ -31,16 +31,35 @@
...
+ */
 int
-azure_connection::list_objects(std::vector<std::string> &objects) const
+azure_connection::list_objects(
+  const std::string &prefix, std::vector<std::string> &objects, bool list_single) const
 {
+    Azure::Storage::Blobs::ListBlobsOptions blob_parameters;
+    blob_parameters.Prefix = prefix;
+    if (list_single)
+        blob_parameters.PageSizeHint = 1;
```
Comment (korteland) [inline ext/storage_sources/azure_store/azure_connection.cpp:56]: It's not obvious that this is hard upper limit rather than an advisory thing, so a comment to that effect would be useful.

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 12]
```diff
@@ -31,16 +31,35 @@
...
+
+    auto list_blobs_response = _azure_client.ListBlobs(blob_parameters);
+
+    for (auto blob_item : list_blobs_response.Blobs) {
```
Comment (korteland) [inline ext/storage_sources/azure_store/azure_connection.cpp:60]: `for (const auto blob_item...` would be a little more appropriate here.

File: ext/storage_sources/azure_store/azure_connection.h [hunk 13]
```diff
@@ -33,10 +33,20 @@
...
+ * This class represents an active connection to the Azure endpoint and allows for interaction with
+ * the Azure client. The azure_connection allows for the following functionality:
+ * Listing the container contents filtered by a given prefix and output all or output single, Puts a
+ * blob to the cloud, gets the blob from the cloud, and deletes a blob from the cloud. It also can check
+ * for the existance for a unique container or blob. Each azure_connection is associated with a unique
+ * azure_client with it's own unique container.
```
Comment (korteland) [inline ext/storage_sources/azure_store/azure_connection.h:43]: Couple of grammar nitpicks, sorry for being a pain:

* ", Puts a" -> ", puts a"
* "gets the blob" -> "gets a blob"
* "for the existance for" -> "for the existence of" (note two fixes here)
* "with it's own" -> "with its own"

File: ext/storage_sources/azure_store/test/test_azure_connection.cpp (hunk 6 above)
Comment (korteland) [reply ext/storage_sources/azure_store/test/test_azure_connection.cpp:37]: I still don't quite think this is right, we've not added more code in this file but we've got a bunch more includes? We should only include things in the file where they're needed.

**Review by korteland — COMMENTED** [review]
Almost there :-) Nice work so far!

File: ext/storage_sources/azure_store/test/test_azure_connection.cpp (hunk 6 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/test/test_azure_connection.cpp:34]: I had those added temporarily while i was doing some testing, cleaned up to og template :)

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 10 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp]: okies made the change :)

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 11 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp:57]: comment added

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 12 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp]: implemented :)

File: ext/storage_sources/azure_store/azure_connection.h (hunk 13 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.h]: hahahaaa no im really thankful for it :) changes made

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 11 above)
Comment (korteland) [reply ext/storage_sources/azure_store/azure_connection.cpp:58]: Thanks - one nitpick, the comment should start with a capital letter.

File: ext/storage_sources/azure_store/azure_connection.h [hunk 14]
```diff
@@ -33,10 +33,20 @@
 
 #include <string>
 
+/*
+ * In Azure: Buckets are synonymous with Containers, and Objects are synonymous with Blobs. This
+ * class represents an active connection to the Azure endpoint and allows for interaction with the
+ * Azure client. The azure_connection allows for the following functionality: Listing the container
+ * contents filtered by a given prefix and output all or output single, puts a blob to the cloud,
+ * gets a blob from the cloud, and deletes a blob from the cloud. It also can check for the
+ * existance of a unique container or blob. Each azure_connection is associated with a unique
```
Comment (korteland) [inline ext/storage_sources/azure_store/azure_connection.h:42]: "existance" is still spelt incorrectly - thanks for fixing the rest of it!

File: ext/storage_sources/azure_store/test/test_azure_connection.cpp (hunk 6 above)
Comment (korteland) [reply ext/storage_sources/azure_store/test/test_azure_connection.cpp:34]: Maybe there's something I'm missing, but there are still a bunch of new includes in this patch.

Comment (mariammojid) [reply ext/storage_sources/azure_store/test/test_azure_connection.cpp:34]: cleaned up now tyty :)

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 11 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp:58]: fixed :)

File: ext/storage_sources/azure_store/azure_connection.h (hunk 14 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.h:42]: fixed :) thanks for picking this up

**Review by korteland — APPROVED** [review]
LGTM!

File: ext/storage_sources/azure_store/test/test_azure_connection.cpp [hunk 15]
```diff
@@ -29,6 +29,4 @@
 #define CATCH_CONFIG_MAIN
 #include <catch2/catch.hpp>
 
-#include "azure_connection.h"
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/test/test_azure_connection.cpp:32]: can we revert this file?

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 16]
```diff
@@ -31,16 +31,37 @@
 #include <azure/core.hpp>
 #include <azure/storage/blobs.hpp>
 
+#include <cstdlib>
+#include <iostream>
+#include <vector>
+#include <string>
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/azure_connection.cpp:37]: Can we clean the includes here? It seems some are already included in the .h file (e.g. strin) and we aren't using iostream, or maybe ctsdlib. I think vector might be already included in the .h file somewhere as well.

File: ext/storage_sources/azure_store/azure_connection.h [hunk 17]
```diff
@@ -33,10 +33,20 @@
 
 #include <string>
 
+/*
+ * In Azure: Buckets are synonymous with Containers, and Objects are synonymous with Blobs. This
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/azure_connection.h:37]: Great comment!! I would just re-arrange it the first two sentences such that:
```
* This class represents an active connection to the Azure endpoint and allows for interaction with the
* Azure client. The Azure cloud stroage names buckets as containers and objects as blobs.
```

File: ext/storage_sources/azure_store/test/test_azure_connection.cpp (hunk 15 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/test/test_azure_connection.cpp:32]: okies done :)

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 16 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp:37]: yep done :)

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 18]
```diff
@@ -31,16 +31,37 @@
...
+
 azure_connection::azure_connection(const std::string &bucket_name, const std::string &obj_prefix)
     : _azure_client(Azure::Storage::Blobs::BlobContainerClient::CreateFromConnectionString(
         std::getenv("AZURE_STORAGE_CONNECTION_STRING"), bucket_name)),
       _bucket_name(bucket_name), _object_prefix(obj_prefix)
 {
 }
 
+/*
+ * Build a list of all of the objects in the bucket.
```
[the remark this answers is not in the document]
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.cpp:47]: fixed :)

File: ext/storage_sources/azure_store/azure_connection.h (hunk 17 above)
Comment (mariammojid) [reply ext/storage_sources/azure_store/azure_connection.h:37]: done :)

**Review by jiechenbo — APPROVED** [review]
LGTM!
